dirty_cat/super_vectorizer.py

```diff
@@ -0,0 +1,341 @@
+"""
+
+This class implements the SuperVectorizer, which is a preprocessor used to
+automatically apply encoders to different types of data, without the need to
+manually categorize them beforehand, or construct complex Pipelines.
+
+"""
+
+# Author: Lilian Boulard <lilian@boulard.fr> | https://github.com/LilianBoulard
+
+import numpy as np
+import pandas as pd
+
+from sklearn.base import BaseEstimator
+from sklearn.compose import ColumnTransformer
+from sklearn.preprocessing import OneHotEncoder
+
+from dirty_cat import SimilarityEncoder
+
+
+_ERR_MSG_UNSUPPORTED_ARR_T = 'Unsupported array type: {}'
+_ERR_MSG_FOUND_NAN = 'Found NaN in array'
+_ERR_MSG_FOUND_INF = 'Found INF in array'
+
+
+def _has_missing_values(array) -> bool:
+    """
+    Returns True if `array` contains missing values, False otherwise.
+    """
+    if isinstance(array, pd.DataFrame) or isinstance(array, pd.Series):
+        return any(array.isnull())
+    elif isinstance(array, np.ndarray):
+        return np.isnan(array).any()
+    else:
+        raise ValueError(_ERR_MSG_UNSUPPORTED_ARR_T.format(type(array)))
+
+
+def _replace_missing(array):
+    """
+    Takes an array, replaces the missing values with a specific value, and returns it.
+    """
+    if isinstance(array, pd.Series):
+        dtype_name = array.dtype.name
+        if dtype_name == 'category' \
+                and ('' not in array.cat.categories):
+            array = array.cat.add_categories('')
+        if dtype_name.startswith('int') or dtype_name.startswith('float'):
+            array.fillna(0, inplace=True)
+        else:
+            array.fillna('', inplace=True)
+        array = array.reset_index(drop=True)
+        return array
+    elif isinstance(array, pd.DataFrame):
+        # Replace missing values for pandas
+        for col in array.columns:
+            dtype_name = array[col].dtype.name
+            if dtype_name == 'category' \
+                    and ('' not in array[col].cat.categories):
+                array[col].cat.add_categories('', inplace=True)
+            array[col].replace(to_replace='?', value='', inplace=True)
+            if dtype_name.startswith('int') or dtype_name.startswith('float'):
+                array[col].fillna(0, inplace=True)
+            else:
+                array[col].fillna('', inplace=True)
+        array = array.reset_index(drop=True)
+        return array
+    elif isinstance(array, np.ndarray):
+        # Replace missing values for numpy
+        array[np.where(np.isnan(array))] = 0
```

Comment:
Here you are replace the values inplace. This is in general dangerous (though certainly not frowned upon) if the programmer is not aware.

The challenge of this function is that it does a bit of both: it modifies the input data, but what it returns can be different. It's a bit bad style, but I have no proposal that is as memory-friendly :).

This should at least be mentioned clearly in a very short docstring, so that later contributors don't make mistakes.